common/src/main/java/io/pravega/common/io/FileModificationWatcher.java

```diff
@@ -0,0 +1,192 @@
+     *                          is useful for testing only.
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     */
+    @VisibleForTesting
+    FileModificationWatcher(@NonNull Path pathOfFileToWatch, @NonNull Consumer<WatchEvent<?>> callback,
+                                   boolean loopContinuously)
+            throws FileNotFoundException {
+        super();
+
+        this.pathOfFileToWatch = pathOfFileToWatch;
+
+        if (!this.pathOfFileToWatch.toFile().exists()) {
+            throw new FileNotFoundException(String.format("File [%s] does not exist.", pathOfFileToWatch));
+        }
+        this.callback = callback;
+        setUncaughtExceptionHandler(uncaughtExceptionalHandler);
+
+        // Set the name for this object/thread for identification purposes.
+        this.setName("file-update-watcher-" + THREAD_NUM.incrementAndGet());
+        this.loopContinuously = loopContinuously;
+    }
+
+    @VisibleForTesting
+    String nameOfFileToWatch() {
+        Path fileName = this.pathOfFileToWatch.getFileName();
+        if (fileName != null) {
+            return fileName.toString();
+        } else {
+            throw new RuntimeException("File name is null");
+        }
+    }
+
+    @VisibleForTesting
+    Path directoryOfFileToWatch() {
+        assert this.pathOfFileToWatch != null;
+        return this.pathOfFileToWatch.getParent();
+    }
+
+    @Override
+    @SuppressWarnings("SleepWhileInLoop")
+    public void run() {
+        WatchKey watchKey = null;
+        WatchService watchService = null;
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            log.debug("Done creating watch service for watching file at path: {}", this.pathOfFileToWatch);
+
+            String fileName = nameOfFileToWatch();
+            Path directoryPath = directoryOfFileToWatch();
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.pathOfFileToWatch);
+
+            while (true) {
+                watchKey = watchService.take();
+                log.info("Retrieved and removed watch key for watching file at path: {}", this.pathOfFileToWatch);
+
+                // Looks odd, right? Using the logic or de-duplicating file change events, as suggested by some here:
+                // https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-
+                // the-same-event
```

Comment:
`io.pravega.common.Exceptions#handleInterrupted`  can be used.

Reply:
I wanted to log a message when the condition occurs for debugging purposes, so have handled the exception using a catch block.